FSDP/T5_training.py

```diff
@@ -0,0 +1,346 @@
+                batch[key] = batch[key].to(local_rank)
+            output = model(input_ids=batch["source_ids"],attention_mask=batch["source_mask"],labels=batch["target_ids"])
+            fsdp_loss[0] += output["loss"].item()  # sum up batch loss
+            fsdp_loss[1] += len(batch)
+
+            if rank==0:
+                inner_pbar.update(1)
+
+    dist.all_reduce(fsdp_loss, op=dist.ReduceOp.SUM)
+    val_loss = fsdp_loss[0] / fsdp_loss[1]
+    if rank == 0:
+        inner_pbar.close()
+        print(f"Validation Loss: {val_loss:.4f}")
+    return val_loss
+
+
+def setup_model(model_name):
+        model = T5ForConditionalGeneration.from_pretrained(model_name)
+        tokenizer =  T5Tokenizer.from_pretrained(model_name)
+        return model, tokenizer
+
+def fsdp_main(args):
+
+    model, tokenizer = setup_model("t5-base")
+
+    local_rank = int(os.environ['LOCAL_RANK'])
+    rank = int(os.environ['RANK'])
+    world_size = int(os.environ['WORLD_SIZE'])
+
+
+    dataset = load_dataset('wikihow', 'all', data_dir='data/')
+    print(dataset.keys())
+    print("Size of train dataset: ", dataset['train'].shape)
+    print("Size of Validation dataset: ", dataset['validation'].shape)
+
+   
+    #wikihow(tokenizer, type_path, num_samples, input_length, output_length, print_text=False)
+    train_dataset = wikihow(tokenizer, 'train', 1500, 512, 150, False) 
+    val_dataset = wikihow(tokenizer, 'validation', 300, 512, 150, False)
+ 
+    sampler1 = DistributedSampler(train_dataset, rank=rank, num_replicas=world_size, shuffle=True)
+    sampler2 = DistributedSampler(val_dataset, rank=rank, num_replicas=world_size)
+
+    setup()
+
+
+    train_kwargs = {'batch_size': args.batch_size, 'sampler': sampler1}
+    test_kwargs = {'batch_size': args.test_batch_size, 'sampler': sampler2}
+    cuda_kwargs = {'num_workers': 2,
+                    'pin_memory': True,
+                    'shuffle': False}
+    train_kwargs.update(cuda_kwargs)
+    test_kwargs.update(cuda_kwargs)
+
+    train_loader = torch.utils.data.DataLoader(train_dataset,**train_kwargs)
+    val_loader = torch.utils.data.DataLoader(val_dataset, **test_kwargs)
+    
+    t5_auto_wrap_policy = functools.partial(
+        transformer_auto_wrap_policy,
+        transformer_layer_cls={
+            T5Block,
+        },
+    )
+    sharding_strategy: ShardingStrategy = ShardingStrategy.SHARD_GRAD_OP #for Zero2 and FULL_SHARD for Zero3
+    torch.cuda.set_device(local_rank)
+   
+   
+    #init_start_event = torch.cuda.Event(enable_timing=True)
+    #init_end_event = torch.cuda.Event(enable_timing=True)
+
+    #init_start_event.record()
+    
+    bf16_ready = (
+    torch.version.cuda
+    and torch.cuda.is_bf16_supported()
+    and LooseVersion(torch.version.cuda) >= "11.0"
+    and dist.is_nccl_available()
+    and nccl.version() >= (2, 10)
+    )
+    
+    if bf16_ready:
+        mp_policy = bfSixteen
+    else:
+        mp_policy = None # defaults to fp32
```

Comment:
@HamidShojanazeri How about wrapping these lines and line 65-85 into a function? e.g. mp_policy = get_mp_policy(...)

Reply:
I have done this in a different codebase and agree it should be modularized.